### PR TITLE
apiextensions: use etag in openapi overlap protection integration test

### PR DESCRIPTION
**Pull request by roycaihw**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Followup of https://github.com/kubernetes/kubernetes/pull/81436. Expose HTTP response header in rest client. Utilize ETag when downloading OpenAPI spec in integration test.

**Does this PR introduce a user-facing change?**:

```release-note
Add a method to rest client `Result` that exposes the HTTP header of the response.
```

/sig api-machinery
/area custom-resources
/assign @sttts

### CODE REVIEW

**Comment by roycaihw** [comment]
Logged the time spent running the test locally:
```
--- PASS: TestOpenAPIApiextensionsOverlapProtection (9.67s)
E0925 20:21:24.827883  179236 kube_apiserver_test.go:192] >>>>> Time elapsed starting test server: 3.872319409s
E0925 20:21:26.135964  179236 kube_apiserver_test.go:335] > time elapsed creating probe CRD 507.770107ms
E0925 20:21:26.135993  179236 kube_apiserver_test.go:339] >>> starting polling openapi spec with etag: "70AEA8D700F4470C0293822378D855FA552EE1A8D483A1B78DE33773F7C033F380BB46B64729C52F7FC148FBAB92EA68CF1B5DEB334320402F2F26A1679D996A"
E0925 20:21:26.237239  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 1.012177ms
E0925 20:21:26.336951  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 748.071µs
E0925 20:21:26.437002  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 777.03µs
E0925 20:21:26.536885  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 680.003µs
E0925 20:21:26.636995  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 805.61µs
E0925 20:21:26.736902  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 711.311µs
E0925 20:21:26.839467  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 521.68µs
E0925 20:21:26.937098  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 897.36µs
E0925 20:21:27.037177  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 978.987µs
E0925 20:21:27.137042  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 844.433µs
E0925 20:21:28.032461  179236 kube_apiserver_test.go:344] >>> openapi spec modified: true, time elapsed: 796.326942ms
E0925 20:21:28.033427  179236 kube_apiserver_test.go:248] > time elapsed marshalling CRD openapi path: 937.844µs
E0925 20:21:28.297820  179236 kube_apiserver_test.go:259] > time elapsed marshalling the entire openapi spec: 264.379063ms
E0925 20:21:29.311948  179236 kube_apiserver_test.go:335] > time elapsed creating probe CRD 504.833269ms
E0925 20:21:29.311980  179236 kube_apiserver_test.go:339] >>> starting polling openapi spec with etag: "FFC62924A09879592947BC39999B23D587C757E4DFBFF1873F1140F54F19A625D294EDE14BF55F8750EF9FC75E34E3714D9B25111AE0F9A4A66D2662DEA082D1"
E0925 20:21:29.412783  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 637.234µs
E0925 20:21:29.512942  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 632.569µs
E0925 20:21:29.613042  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 897.072µs
E0925 20:21:29.712992  179236 kube_apiserver_test.go:344] >>> openapi spec modified: false, time elapsed: 837.049µs
E0925 20:21:30.623411  179236 kube_apiserver_test.go:344] >>> openapi spec modified: true, time elapsed: 811.217027ms
E0925 20:21:30.623534  179236 kube_apiserver_test.go:305] > time elapsed marshalling CRD openapi definition: 92.473µs
E0925 20:21:30.623546  179236 kube_apiserver_test.go:309] >>>>> Total time elapsed running test logic (excluding server startup time): 5.79563997s
```

File: staging/src/k8s.io/client-go/rest/request.go [hunk 1]
```diff
@@ -1114,6 +1118,12 @@ func (r Result) StatusCode(statusCode *int) Result {
 	return r
 }
 
+// Header returns the HTTP header of the response.
```
Comment (jpbetz) [inline staging/src/k8s.io/client-go/rest/request.go]: nit: Function doesn't return the header, although I do see the above godoc for StatusCode is also phrased this way. Change both to a form like "Header sets the provided header to the HTTP header of the response"?

**Review by jpbetz — COMMENTED** [review]
Just one nit, looks good otherwise

File: staging/src/k8s.io/client-go/rest/request.go (hunk 1 above)
Comment (roycaihw) [reply staging/src/k8s.io/client-go/rest/request.go]: thanks. Updated

**Review by jpbetz — COMMENTED** [review]
/lgtm

**Comment by roycaihw** [comment]
/assign @sttts 
for approval

**Pull request closed without merging**